### PR TITLE
Decrement the counter only if the connection is removed

### PR DESCRIPTION
**Pull request by gburanov**
https://trello.com/c/6aveL45w/3246-debug-forwarding-only-works-once

What happens:
* We actually open connection only once
* But do write in goroutine, this can lead to removing the bad connection several times (once for each write)
* This is fine for the map, but not fine for our counter (that goes over 0 to very high value - because it is unsigned int)
* Because of that - we will have following if failing each time https://github.com/remerge/go-service/blob/master/debug_forwarder.go#L67

Solution is to use mutext - IMHO it is fine, because it it goroutine

### CODE REVIEW

File: debug_forwarder.go [hunk 1]
```diff
@@ -123,9 +125,14 @@ func (f *DebugForwarder) Write(data []byte) (n int, err error) {
 				}
 			}
 			if badConn {
-				f.conns.Delete(k)
-				atomic.AddUint32(&f.connCount, ^uint32(0))
-				f.connCloseCounter.Inc(1)
+				f.badConnectionMutex.Lock()
+				defer f.badConnectionMutex.Unlock()
```
Comment (gburanov) [inline debug_forwarder.go]: I think mutex in goroutine is fine, but u can suggest other options =)

Comment (akaspin) [reply debug_forwarder.go]:
```
// use this instead naked connection
type conn struct {
guard uint32

...
}

func (c *conn) Close() error {
if atomic.CompareAndSwapUint32(&c.guard, 1) {
// close and *decrement* f.connCount here
}
return nil
}
```

Comment (gburanov) [reply debug_forwarder.go]: yes, u are right, it seems to be better approach

File: debug_forwarder.go [hunk 2]
```diff
@@ -34,6 +34,19 @@ type DebugForwarder struct {
...
+	f     *DebugForwarder
+	guard uint32
+}
+
+func (c *connection) Close() error {
+	if atomic.CompareAndSwapUint32(&c.guard, 0, 1) {
+		atomic.AddUint32(&c.f.connCount, ^uint32(0))
+		c.f.connCloseCounter.Inc(1)
+	}
+	return c.Conn.Close()
```
Comment (akaspin) [inline debug_forwarder.go]: @gburanov It's better to move Close call inside protected block.

**Pull request merged**